File: test-app/frontend/src/api/api-watch.ts
// Watch API

import { GetAPIURL, type RequestParams } from "@/utils/request";

export interface SubStream {
    width: number;
    height: number;
    fps: number;
    indexFile: string;
}

export interface ChannelStatus {
    id: string;
    record: boolean;
    resolutions: string;
    previews: string;
    live: boolean;
    streamId: string;
    liveStartTimestamp: number;
    liveSubStreams: SubStream[];
}

export interface VODItem {
    streamId: string;
    timestamp: string;
}

export interface VODItemList {
    vod_list: VODItem[];
}

export interface VODStreaming {
    streamId: string;
    timestamp: number;
    subStreams: SubStream[];
    hasPreviews: boolean;
    previewsIndex: string;
}

export class WatchAPI {
    public static GetChannelStatus(channel: string): RequestParams {
        return {
            method: "GET",
            url: GetAPIURL("/api/watch/" + encodeURIComponent(channel)),
        };
    }

    public static GetChannelVODList(channel: string): RequestParams {
        return {
            method: "GET",
            url: GetAPIURL("/api/watch/" + encodeURIComponent(channel) + "/vod"),
        };
    }

    public static GetChannelVOD(channel: string, streamId: string): RequestParams {
        return {
            method: "GET",
            url: GetAPIURL("/api/watch/" + encodeURIComponent(channel) + "/vod/" + encodeURIComponent(streamId)),
        };
    }

    public static CreateAccount(username: string, password: string, write: boolean): RequestParams {
        return {
            method: "POST",
            url: GetAPIURL("/api/admin/accounts"),
            json: {
                username: username,
                password: password,
                write: write,
            },
        };
    }

    public static DeleteAccount(username: string): RequestParams {
        return {
            method: "POST",
            url: GetAPIURL("/api/admin/accounts/delete"),
            json: {
                username: username,
            },
        };
    }
}

File: test-app/backend/api_watch.go
// Watch API

package main

import (
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
)

type ChannelStatusResponse struct {
	Id string `json:"id"`

	Record      bool   `json:"record"`
	Resolutions string `json:"resolutions"`
	Previews    string `json:"previews"`

	Live     bool   `json:"live"`
	StreamId string `json:"streamId"`

	LiveStartTimestamp int64 `json:"liveStartTimestamp"`

	LiveSubStreams []SubStreamWithCdn `json:"liveSubStreams"`
}

func api_getChannelStatus(response http.ResponseWriter, request *http.Request) {
	vars := mux.Vars(request)

	channel := vars["channel"]

	if !validateStreamIDString(channel) {
		ReturnAPIError(response, 400, "BAD_REQUEST", "Bad request.")
		return
	}

	result := DATABASE.GetChannelStatus(channel)

	if result == nil {
		ReturnAPIError(response, 404, "NOT_FOUND", "Channel not found.")
		return
	}

	jsonResult, err := json.Marshal(result)

	if err != nil {
		LogError(err)

		ReturnAPIError(response, 500, "INTERNAL_ERROR", "Internal server error, Check the logs for details.")
		return
	}

	ReturnAPI_JSON(response, request, jsonResult)
}

type VODItem struct {
	StreamId  string `json:"streamId"`
	Timestamp int64  `json:"timestamp"`
}

type ListVODResponse struct {
	VODS []VODItem `json:"vod_list"`
}

func api_listVODs(response http.ResponseWriter, request *http.Request) {
	vars := mux.Vars(request)

	channel := vars["channel"]

	if !validateStreamIDString(channel) {
		ReturnAPIError(response, 400, "BAD_REQUEST", "Bad request.")
		return
	}

	result := ListVODResponse{
		VODS: DATABASE.GetChannelVODList(channel),
	}

	if result.VODS == nil {
		ReturnAPIError(response, 404, "NOT_FOUND", "Channel not found.")
		return
	}

	jsonResult, err := json.Marshal(result)

	if err != nil {
		LogError(err)

		ReturnAPIError(response, 500, "INTERNAL_ERROR", "Internal server error, Check the logs for details.")
		return
	}

	ReturnAPI_JSON(response, request, jsonResult)
}

func api_getVOD(response http.ResponseWriter, request *http.Request) {
	vars := mux.Vars(request)

	channel := vars["channel"]
	streamId := vars["vod"]

	if !validateStreamIDString(channel) || !validateStreamIDString(streamId) {
		ReturnAPIError(response, 400, "BAD_REQUEST", "Bad request.")
		return
	}

	result := DATABASE.GetChannelVOD(channel, streamId)

	if result == nil {
		ReturnAPIError(response, 404, "NOT_FOUND", "Channel not found.")
		return
	}

	jsonResult, err := json.Marshal(result)

	if err != nil {
		LogError(err)

		ReturnAPIError(response, 500, "INTERNAL_ERROR", "Internal server error, Check the logs for details.")
		return
	}

	ReturnAPI_JSON(response, request, jsonResult)
}
